File: kongen_router.js
/**
 * Kongen Smart Router — Route LLM calls through Kongen for 60% cost savings.
 *
 * Scores each prompt with Kongen's morphogenetic analysis to determine complexity,
 * then routes to the cheapest model that can handle it.
 *
 * Usage:
 *   npm install @anthropic-ai/sdk openai
 *
 *   export KONGEN_API_KEY="kl_live_..."
 *   export ANTHROPIC_API_KEY="sk-ant-..."   # for Claude
 *   export OPENAI_API_KEY="sk-..."          # for OpenAI
 *
 *   node kongen_router.js
 */

const KONGEN_BASE =
  process.env.KONGEN_API_BASE_URL || "https://api.kongenlabs.life";
const KONGEN_KEY = process.env.KONGEN_API_KEY;

// --- Model tiers (cheapest → most capable) ---

const CLAUDE_TIERS = [
  { name: "claude-haiku-4-5-20251001", inputCost: 1.0, outputCost: 5.0, maxRegime: "fast" },
  { name: "claude-sonnet-4-6", inputCost: 3.0, outputCost: 15.0, maxRegime: "deep" },
  { name: "claude-opus-4-6", inputCost: 5.0, outputCost: 25.0, maxRegime: "exhaustive" },
];

const OPENAI_TIERS = [
  { name: "gpt-4o-mini", inputCost: 0.15, outputCost: 0.60, maxRegime: "fast" },
  { name: "gpt-4o", inputCost: 2.50, outputCost: 10.0, maxRegime: "deep" },
  { name: "o3", inputCost: 10.0, outputCost: 40.0, maxRegime: "exhaustive" },
];

const REGIME_ORDER = ["trivial", "fast", "moderate", "deep", "exhaustive"];

function pickModel(regime, tiers) {
  const idx = REGIME_ORDER.indexOf(regime);
  const regimeIdx = idx >= 0 ? idx : 2;
  for (const tier of tiers) {
    if (REGIME_ORDER.indexOf(tier.maxRegime) >= regimeIdx) return tier;
  }
  return tiers[tiers.length - 1];
}

function estimateCost(model, inputTokens, outputTokens) {
  return (
    (inputTokens * model.inputCost) / 1_000_000 +
    (outputTokens * model.outputCost) / 1_000_000
  );
}

// --- Kongen API call ---

async function scorePrompt(text) {
  const res = await fetch(`${KONGEN_BASE}/v1/logic/score`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      "X-API-Key": KONGEN_KEY,
    },
    body: JSON.stringify({ text }),
  });
  if (!res.ok) {
    const err = await res.json().catch(() => ({}));
    throw new Error(`Kongen API error ${res.status}: ${err.detail || res.statusText}`);
  }
  return res.json();
}

// --- Smart Router ---

class KongenRouter {
  constructor(provider = "claude") {
    this.provider = provider;
    this.tiers = provider === "claude" ? CLAUDE_TIERS : OPENAI_TIERS;
    this.totalSpent = 0;
    this.totalSaved = 0;
    this.calls = 0;
  }

  async complete(prompt, { system = "You are a helpful assistant.", maxTokens } = {}) {
    // 1. Score with Kongen (1 KT)
    const score = await scorePrompt(prompt);
    const regime = score.regime;
    const recommended = score.recommended_tokens;

    // 2. Pick cheapest capable model
    const model = pickModel(regime, this.tiers);
    const expensive = this.tiers[this.tiers.length - 1];

    // 3. Use recommended tokens if no override
    const tokens = maxTokens || Math.min(recommended, 4096);

    // 4. Call the LLM
    let text, inputTokens, outputTokens;

    if (this.provider === "claude") {
      const Anthropic = (await import("@anthropic-ai/sdk")).default;
      const client = new Anthropic();
      const response = await client.messages.create({
        model: model.name,
        max_tokens: tokens,
        system,
        messages: [{ role: "user", content: prompt }],
      });
      text = response.content[0].text;
      inputTokens = response.usage.input_tokens;
      outputTokens = response.usage.output_tokens;
    } else {
      const OpenAI = (await import("openai")).default;
      const client = new OpenAI();
      const response = await client.chat.completions.create({
        model: model.name,
        max_tokens: tokens,
        messages: [
          { role: "system", content: system },
          { role: "user", content: prompt },
        ],
      });
      text = response.choices[0].message.content;
      inputTokens = response.usage.prompt_tokens;
      outputTokens = response.usage.completion_tokens;
    }

    // 5. Calculate savings
    const cost = estimateCost(model, inputTokens, outputTokens);
    const wouldCost = estimateCost(expensive, inputTokens, outputTokens);
    const saved = wouldCost - cost;

    this.totalSpent += cost;
    this.totalSaved += saved;
    this.calls++;

    return {
      text,
      model: model.name,
      regime,
      boostFactor: score.boost_factor,
      cost,
      wouldCost,
      saved,
      inputTokens,
      outputTokens,
    };
  }

  stats() {
    const total = this.totalSpent + this.totalSaved;
    return {
      calls: this.calls,
      totalSpent: +this.totalSpent.toFixed(4),
      totalSaved: +this.totalSaved.toFixed(4),
      savingsPct: total > 0 ? +((this.totalSaved / total) * 100).toFixed(1) : 0,
    };
  }
}

// --- Demo ---

async function main() {
  if (!KONGEN_KEY) {
    console.log("Set KONGEN_API_KEY (get one at https://garden.kongenlabs.life/keys)");
    console.log("Also set ANTHROPIC_API_KEY or OPENAI_API_KEY");
    process.exit(1);
  }

  const provider = process.env.ANTHROPIC_API_KEY ? "claude" : "openai";
  const router = new KongenRouter(provider);
  console.log(`Kongen Smart Router (${provider})\n${"=".repeat(50)}\n`);

  const prompts = [
    "What is 2 + 2?",
    "Summarize the key differences between TCP and UDP.",
    "Write a Python function to check if a number is prime.",
    "Explain the implications of Godel's incompleteness theorems on AI.",
    "Prove that the sum of the first n odd numbers equals n squared.",
  ];

  for (const prompt of prompts) {
    const r = await router.complete(prompt);
    console.log(`Prompt: ${prompt.slice(0, 60)}...`);
    console.log(`  Regime: ${r.regime} -> Model: ${r.model}`);
    console.log(`  Cost: $${r.cost.toFixed(4)} (would be $${r.wouldCost.toFixed(4)}, saved $${r.saved.toFixed(4)})`);
    console.log(`  Tokens: ${r.inputTokens}in/${r.outputTokens}out\n`);
  }

  const s = router.stats();
  console.log("=".repeat(50));
  console.log(
    `Total: ${s.calls} calls, $${s.totalSpent} spent, $${s.totalSaved} saved (${s.savingsPct}%)`
  );
}

main().catch(console.error);

File: kongen_router.py
"""
Kongen Smart Router — Route LLM calls through Kongen for 60% cost savings.

Scores each prompt with Kongen's morphogenetic analysis to determine complexity,
then routes to the cheapest model that can handle it. Works with both Anthropic
(Claude) and OpenAI APIs.

Usage:
    pip install kongenlabs anthropic openai

    export KONGEN_API_KEY="kl_live_..."
    export ANTHROPIC_API_KEY="sk-ant-..."   # for Claude
    export OPENAI_API_KEY="sk-..."          # for OpenAI

    python kongen_router.py
"""

from __future__ import annotations

import os
from dataclasses import dataclass

from kongen import KongenClient


# --- Model routing configuration ---

@dataclass
class ModelTier:
    """LLM model tier with cost info."""
    name: str
    input_cost_per_m: float   # $ per 1M input tokens
    output_cost_per_m: float  # $ per 1M output tokens
    max_regime: str           # highest regime this model handles well


# Claude model tiers (cheapest → most capable)
CLAUDE_TIERS = [
    ModelTier("claude-haiku-4-5-20251001", 1.0, 5.0, "fast"),
    ModelTier("claude-sonnet-4-6", 3.0, 15.0, "deep"),
    ModelTier("claude-opus-4-6", 5.0, 25.0, "exhaustive"),
]

# OpenAI model tiers
OPENAI_TIERS = [
    ModelTier("gpt-4o-mini", 0.15, 0.60, "fast"),
    ModelTier("gpt-4o", 2.50, 10.0, "deep"),
    ModelTier("o3", 10.0, 40.0, "exhaustive"),
]

# Regime ordering (simplest → most complex)
REGIME_ORDER = ["trivial", "fast", "moderate", "deep", "exhaustive"]


def pick_model(regime: str, tiers: list[ModelTier]) -> ModelTier:
    """Pick the cheapest model that can handle the given regime."""
    regime_idx = REGIME_ORDER.index(regime) if regime in REGIME_ORDER else 2
    for tier in tiers:
        tier_idx = REGIME_ORDER.index(tier.max_regime)
        if tier_idx >= regime_idx:
            return tier
    return tiers[-1]  # fallback to most capable


def estimate_cost(model: ModelTier, input_tokens: int, output_tokens: int) -> float:
    """Estimate cost in dollars."""
    return (
        input_tokens * model.input_cost_per_m / 1_000_000
        + output_tokens * model.output_cost_per_m / 1_000_000
    )


# --- Smart Router ---

class KongenRouter:
    """
    Routes prompts to the optimal LLM model using Kongen's pattern analysis.

    Kongen scores each prompt to determine its reasoning complexity (regime),
    then the router picks the cheapest model that handles that regime.

    Example:
        router = KongenRouter(provider="claude")
        response = router.complete("What is 2+2?")
        # → Routes to Haiku (cheapest), saves ~80% vs always using Opus

        response = router.complete("Prove that P != NP")
        # → Routes to Opus (most capable), no savings but correct model
    """

    def __init__(
        self,
        provider: str = "claude",
        kongen_api_key: str | None = None,
        llm_api_key: str | None = None,
    ):
        self.provider = provider
        self.kongen = KongenClient(api_key=kongen_api_key)

        if provider == "claude":
            import anthropic
            self.llm = anthropic.Anthropic(api_key=llm_api_key)
            self.tiers = CLAUDE_TIERS
        elif provider == "openai":
            import openai
            self.llm = openai.OpenAI(api_key=llm_api_key)
            self.tiers = OPENAI_TIERS
        else:
            raise ValueError(f"Unknown provider: {provider}. Use 'claude' or 'openai'.")

        self.total_saved = 0.0
        self.total_spent = 0.0
        self.calls = 0

    def complete(
        self,
        prompt: str,
        system: str = "You are a helpful assistant.",
        max_tokens: int | None = None,
    ) -> dict:
        """
        Score prompt with Kongen, route to optimal model, return response.

        Returns dict with: text, model, regime, cost, would_cost, saved
        """
        # 1. Score with Kongen (1 KT)
        score = self.kongen.chiryu.score(prompt)
        regime = score.regime
        recommended = score.recommended_tokens

        # 2. Pick cheapest capable model
        model = pick_model(regime, self.tiers)
        expensive = self.tiers[-1]  # what you'd use without routing

        # 3. Use recommended tokens if no override
        if max_tokens is None:
            max_tokens = min(recommended, 4096)

        # 4. Call the LLM
        if self.provider == "claude":
            response = self.llm.messages.create(
                model=model.name,
                max_tokens=max_tokens,
                system=system,
                messages=[{"role": "user", "content": prompt}],
            )
            text = response.content[0].text
            input_tokens = response.usage.input_tokens
            output_tokens = response.usage.output_tokens
        else:
            response = self.llm.chat.completions.create(
                model=model.name,
                max_tokens=max_tokens,
                messages=[
                    {"role": "system", "content": system},
                    {"role": "user", "content": prompt},
                ],
            )
            text = response.choices[0].message.content
            input_tokens = response.usage.prompt_tokens
            output_tokens = response.usage.completion_tokens

        # 5. Calculate savings
        actual_cost = estimate_cost(model, input_tokens, output_tokens)
        would_cost = estimate_cost(expensive, input_tokens, output_tokens)
        saved = would_cost - actual_cost

        self.total_spent += actual_cost
        self.total_saved += saved
        self.calls += 1

        return {
            "text": text,
            "model": model.name,
            "regime": regime,
            "boost_factor": score.boost_factor,
            "a_i_ratio": score.a_i_ratio,
            "cost": actual_cost,
            "would_cost": would_cost,
            "saved": saved,
            "input_tokens": input_tokens,
            "output_tokens": output_tokens,
        }

    def stats(self) -> dict:
        """Return cumulative routing statistics."""
        return {
            "calls": self.calls,
            "total_spent": round(self.total_spent, 4),
            "total_saved": round(self.total_saved, 4),
            "savings_pct": round(
                self.total_saved / (self.total_spent + self.total_saved) * 100, 1
            ) if (self.total_spent + self.total_saved) > 0 else 0,
        }


# --- Demo ---

if __name__ == "__main__":
    # Pick provider based on which API key is available
    if os.environ.get("ANTHROPIC_API_KEY"):
        provider = "claude"
    elif os.environ.get("OPENAI_API_KEY"):
        provider = "openai"
    else:
        print("Set ANTHROPIC_API_KEY or OPENAI_API_KEY to run the demo.")
        print("Also set KONGEN_API_KEY (get one at https://garden.kongenlabs.life/keys)")
        raise SystemExit(1)

    router = KongenRouter(provider=provider)
    print(f"Kongen Smart Router ({provider})\n{'=' * 50}\n")

    # Mix of simple and complex prompts
    prompts = [
        "What is 2 + 2?",
        "Summarize the key differences between TCP and UDP.",
        "Write a Python function to check if a number is prime.",
        "Explain the implications of Godel's incompleteness theorems on AI.",
        "Prove that the sum of the first n odd numbers equals n squared.",
    ]

    for prompt in prompts:
        result = router.complete(prompt)
        print(f"Prompt: {prompt[:60]}...")
        print(f"  Regime: {result['regime']} -> Model: {result['model']}")
        print(f"  Cost: ${result['cost']:.4f} (would be ${result['would_cost']:.4f}, saved ${result['saved']:.4f})")
        print(f"  Tokens: {result['input_tokens']}in/{result['output_tokens']}out")
        print()

    s = router.stats()
    print(f"{'=' * 50}")
    print(f"Total: {s['calls']} calls, ${s['total_spent']:.4f} spent, ${s['total_saved']:.4f} saved ({s['savings_pct']}%)")
